### ISSUE TITLE
multiobjective_optimization.ipynb ex. could be run in parallel not noisy ? 

### ISSUE THREAD

**Issue by hanochk** [facebook/Ax#731]
Hi as for the example multiobjective_optimization.ipynb : it runs very slow over my problem. How do I get it working with a number of arms to run a batch in parallel? Using get_MOO_EHVI() means it isn't the q version qEHVI (which should be the parallel version)?  
Suppose my optimization isn't based on noisy samples, for having the q version do I have only the get_MOO_NEHVI() which is related to noisy examples? 
How do I get verbose printing of the optimization by get_MOO_EHVI() in that example like getting verbose in the saasbo_nehvi.ipynb (though they are different)?
Thanks, Hanoch

**Comment by sdaulton** [comment]
> it runs very slow over my problem

How many objectives are you optimizing and how many outcomes are you modeling?

> Using get_MOO_EHVI() means it isn't the q version qEHVI

It actually is "the q-version".  But if you are generating batches of arms, you should use qNEHVI, which is mathematically equivalent even in the noiseless setting, but much faster to compute. I would always recommend using get_MOO_NEHVI.

More generally, we always use the Monte-Carlo acquisition functions (denoted by q).

> How do I get verbose printing of the optimization by get_MOO_EHVI() in that example like getting verbose in the saasbo_nehvi.ipynb (though they are different)?

The `verbose` setting and printing in saasbo_nehvi.ipynb is for pyro's NUTS implementation. Is is only applicable when using NUTS for model fitting (e.g. in SAASBO).

**Comment by hanochk** [comment]
Hi Sam , thanks for the quick reply. 2 objectives each predicting a scalar.
Does it  related the outcomes?

**Comment by sdaulton** [comment]
No, that should be quite fast. How large is your search space?

**Comment by hanochk** [comment]
The space is [0 1] for 9 dimentions. What is referred as outcome ?

**Comment by hanochk** [comment]
Hi Sam, the main reason for that notebook running slow is the fact that :ehvi_model.gen(1) i.e it doesn't work in parallel. I've replaced the get_MOO_EHVI with get_MOO_NEHVI but the SOBOL loop for instance runs for 30 min at least. What could be done?

**Comment by lena-kashtelyan** [comment]
@hanochk, an outcome is each metric being measured, data for which is being used by the model to generate further candidates.

> the SOBOL loop for instance runs for 30 min at least

That seems quite strange, my guess is that it has to do with fitting the MOO model and hypervolume computation at each iteration. Do you want hypervolume reported at each iteration? If not, you can skip it during the Sobol loop, and it should be very fast.

> how to add more arms for the parallelism

You can reconfigure the code in that notebook to be like this (showing for one of the loops):
```
from ax.modelbridge.modelbridge_utils import get_pending_observation_features
ehvi_hv_list = []
ehvi_model = None

for i in range(N_BATCH / 5):   
    for j in range(5):  # Make 5 trials without re-creating new model (showing sample parallelism of 5)
        trials = []
        ehvi_model = get_MOO_EHVI(
            experiment=ehvi_experiment, 
            data=ehvi_data,
        )
        # Passing pending observations is crucial, otherwise the same trial might be generated over and over.
        generator_run = ehvi_model.gen(1, pending_observations=get_pending_observation_features(ehvi_experiment))
        trial = ehvi_experiment.new_trial(generator_run=generator_run)
        trial.run()
        trials.append(trial)
    
    ehvi_data = Data.from_multiple_data(
        [ehvi_data] + ehvi_experiment.fetch_trials_data([t.index for t in trials)]
    )    
    try:
        hv = observed_hypervolume(modelbridge=ehvi_model)
    except:
        hv = 0
        print("Failed to compute hv")
    print(f"Iteration: {i}, HV: {hv}")
```

Also, at the top of this tutorial is an example of using Ax Service API for multi-objective optimization, which you might find quite a bit easier to use than the Developer API you are using right now: https://ax.dev/tutorials/multiobjective_optimization.html#Using-the-Service-API. There, you can parallelize trials pretty naturally by just generating a few at a time and then completing the few you generated at once. To obtain current hypervolume, you could do `observed_hypervolume(modelbridge=ax_client.generation_strategy.model)`.

**Comment by hanochk** [comment]
Hi Lena, thanks for the explanation.
One of my objectives out of the 2 in the MOO settings, applying inference
given few images (vision optimization), it may explain the 30min of SOBOL.
The HV reporting isn't expected to ease the loop so much in that case,
right?
Furthermore, i was trying to imply my MOO problem over the SAASBO ipynb,
though the objective isn't noisy in my case. Getting the following warning
about sequential, is it normal?

from iteration #19 of SOBOL under saasbo_nehvi.ipynb  :



Unknown solver options: sequential

[WARNING 11-30 11:04:24] ax.service.utils.report_utils: Ignoring
user-specified `deduplicate_on_map_keys = True` since
`exp.fetch_data().map_keys` is empty or does not exist. Check that at least
one element of `metrics` (or `exp.metrics` if `metrics is None`) inherits
from `MapMetric`.
Iteration: 9, HV: 0.0
/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/ans_compression_opt/lib/python3.8/site-packages/botorch/generation/gen.py:166:
OptimizeWarning:

Unknown solver options: sequential


Best regards,
Hanoch Kremer

**Comment by lena-kashtelyan** [comment]
For the solver error, would you mind opening a separate issue, @hanochk?

**Comment by lena-kashtelyan** [comment]
> One of my objectives out of the 2 in the MOO settings, applying inference given few images (vision optimization), it may explain the 30min of SOBOL. The HV reporting isn't expected to ease the loop so much in that case,
right?

I'm not sure I fully understand. Is the question here whether HV reporting is likely accounting for the length of the loop given that evaluating each trial separately is also slow? 

If so, the answer is that it's probably not accounting for that in that case, but to speed up the whole thing you could parallelize trial evaluations into batches of trials as suggested in my comment above: https://github.com/facebook/Ax/issues/731#issuecomment-982143766.

**Comment by hanochk** [comment]
Hi Lena, does the piece of code you added, parallelize the MOBO search relying on the assumption that my customize objectives could be run in parallel in nature? For instance one of the objectives, "compression " is computed via C code under the hood of the python wrapper. Does it impose restrictions in terms of thread-safe or so? or does the MOBO engine (get_MOO_EHVI) can handle that under the hood ?

**Comment by lena-kashtelyan** [comment]
Hi @hanochk,

> does the piece of code you added, parallelize the MOBO search relying on the assumption that my customize objectives could be run in parallel in nature?

It does, yes.

> For instance one of the objectives, "compression " is computed via C code under the hood of the python wrapper. Does it impose restrictions in terms of thread-safe or so? 

That I'm not sure of since I have not seen that code, you might need to figure out that part separately.

> or does the MOBO engine (get_MOO_EHVI) can handle that under the hood?

Here the question is whether your C code for the compression objective could interfere with the one in `gen_MOO_EHVI`, correct? Let me loop in @Balandat and @sdaulton to answer that one. I think it probably wouldn't, but I'm not 100% sure as this is not in Ax but in BoTorch land.

**Comment by hanochk** [comment]
BTW that piece of code doesn't work well after a while: can you help ?

/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/ans_compression_opt/lib/python3.8/site-packages/botorch/generation/gen.py:166:
OptimizeWarning:

Unknown solver options: sequential

/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/ans_compression_opt/lib/python3.8/site-packages/botorch/generation/gen.py:166:
OptimizeWarning:

Unknown solver options: sequential

Traceback (most recent call last):
  File
"/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/HMS/mobo_ANS.py",
line 493, in <module>
    if __name__ == "__main__":
  File
"/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/HMS/mobo_ANS.py",
line 290, in main
    trials.append(trial)
TypeError: can only concatenate list (not "Data") to list

**Comment by hanochk** [comment]
Furthermore it fails from some reason : 
/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/ans_compression_opt/lib/python3.8/site-packages/botorch/generation/gen.py:166: OptimizeWarning:

Unknown solver options: sequential

/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/ans_compression_opt/lib/python3.8/site-packages/botorch/generation/gen.py:166: OptimizeWarning:

Unknown solver options: sequential

Traceback (most recent call last):
  File "/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/HMS/mobo_ANS.py", line 493, in <module>
    if __name__ == "__main__":
  File "/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/HMS/mobo_ANS.py", line 290, in main
    trials.append(trial)
TypeError: can only concatenate list (not "Data") to list

**Comment by sdaulton** [comment]
> does the piece of code you added, parallelize the MOBO search relying on the assumption that my customize objectives could be run in parallel in nature? For instance one of the objectives, "compression " is computed via C code under the hood of the python wrapper. Does it impose restrictions in terms of thread-safe or so? or does the MOBO engine (get_MOO_EHVI) can handle that under the hood ?

For parallel bayesian optimization, all Ax does is generate and return multiple candidates designs. How those candidate designs are evaluated is up to you. You can evaluate them sequentially (repeated calling your objective on each design and waiting for each design to complete before evaluating the next design in the batch) or in parallel (evaluating the objective simultaneously using e.g. multiple processes or machines). If you evaluate objectives in parallel, it is up to you to make sure that evaluating the objective on one design does not corrupt the evaluation of another design (e.g. due to some race condition).

**Comment by lena-kashtelyan** [comment]
> TypeError: can only concatenate list (not "Data") to list

@hanochk, sorry there was a bug in my code snippet. Updated it in the comment above; `trials = []` line needed to be outside of the inner loop (iterating on `j`).

**Comment by hanochk** [comment]
Thanks !

**Comment by hanochk** [comment]
Ok thanks !

**Comment by hanochk** [comment]
What is the reason of the following warnings?

Unknown solver options: sequential

**Comment by hanochk** [comment]
> > does the piece of code you added, parallelize the MOBO search relying on the assumption that my customize objectives could be run in parallel in nature? For instance one of the objectives, "compression " is computed via C code under the hood of the python wrapper. Does it impose restrictions in terms of thread-safe or so? or does the MOBO engine (get_MOO_EHVI) can handle that under the hood ?
> 
> For parallel bayesian optimization, all Ax does is generate and return multiple candidates designs. How those candidate designs are evaluated is up to you. You can evaluate them sequentially (repeated calling your objective on each design and waiting for each design to complete before evaluating the next design in the batch) or in parallel (evaluating the objective simultaneously using e.g. multiple processes or machines). If you evaluate objectives in parallel, it is up to you to make sure that evaluating the objective on one design does not corrupt the evaluation of another design (e.g. due to some race condition).

How does the MOBO framework in the ipynb scales to multi GPU environment ?

**Comment by Balandat** [comment]
> How does the MOBO framework in the ipynb scales to multi GPU environment ?

For the Bayesian Optimization algorithm we currently don't have multi-GPU support (this is quite tricky to do). However, that doesn't mean that you can't use multiple GPUs to run your evaluations on if your evaluation function uses a GPU. One thing to potentially be aware of in this case is memory requirements - if you run a BayesOpt loop locally the designated GPU (by default `cuda:0`) will have memory allocated for the MOBO models, so you may want to make sure that running evaluations on that GPU doesn't cause OOMs (which in a multi-GPU environment is easily achieved by using one of the other GPUs).

**Comment by lena-kashtelyan** [comment]
If you have more follow-ups, feel free to reopen this issue, @hanochk.

**Comment by hanochk** [comment]
Hi Lena unfortunately the list couldn't be handled by the
Data.from_multiple_data() by that list of trials any clue what went wrong
with the code you have shared ?

Unknown solver options: sequential

/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/ans_compression_opt/lib/python3.8/site-packages/botorch/generation/gen.py:166:
OptimizeWarning:

Unknown solver options: sequential

Traceback (most recent call last):
  File
"/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/HMS/mobo_ANS.py",
line 534, in <module>
    main()
  File
"/home/hanoch/PycharmProjects/pythonProject/ans_compression_opt/laila/sefi/_skariel/mixed_quant/HMS/mobo_ANS.py",
line 327, in main
    ehvi_data = Data.from_multiple_data([ehvi_data] +
ehvi_experiment.fetch_trials_data([t.index for t in trials]))
TypeError: can only concatenate list (not "Data") to list
[image: image.png]

**Comment by lena-kashtelyan** [comment]
The correct line is: `Data.from_multiple_data([ehvi_data, ehvi_experiment.fetch_trials_data([t.index for t in trials]))`, replace that one and things should work. Fixed it in the original example, too!

**Comment by hanochk** [comment]
Fine, thanks BTW it should be (missing ]):
Data.from_multiple_data([ehvi_data, ehvi_experiment.fetch_trials_data([t.index
for t in trials])*]*)

Best regards,
Hanoch Kremer
https://hanochk.github.io/about
